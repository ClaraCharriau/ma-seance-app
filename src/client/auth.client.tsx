import { User } from '../models/User';
import axios from 'axios';

const user_auth_path = 'http://localhost:7878/auth';
const user_verify_path = 'http://localhost:7878/verify';
const user_sign_in_path = 'http://localhost:7878/sign-in';

export const loginUser = async (email: string, password: string): Promise<User> => {
    return await axios
        .post(user_auth_path, {
            email: email,
            password: password
        })
        .then(response => {
            console.log('user-auth-client-response POST user :', response.data);
            return response.data;
        })
        .catch(error => {
            console.error('Error: could not log user : ' + error.message);
            throw error;
        });
};

export const checkAccountExists = async (email: string): Promise<boolean> => {
    return await axios
        .post(user_verify_path, {
            email: email
        })
        .then(function (response) {
            console.log('user-verify-client-response POST verify :', response.data);
            if (response.data && typeof response.data.exists === 'boolean') {
                return response.data.exists;
            }
            throw Error('Could not parse api /verify response. ');
        })
        .catch(function (error: Error) {
            console.error('Error: could not verify if user exists' + error.message);
            throw error;
        });
};

export const signIn = async (pseudo: string, email: string, password: string): Promise<User> => {
    return await axios
        .post(user_sign_in_path, {
            pseudo: pseudo,
            email: email,
            password: password
        })
        .then(function (response) {
            console.log('user-sign-in-client-response POST user :', response.data);
            return response.data;
        })
        .catch(function (error: Error) {
            console.error('Error: could not sign in user' + error.message);
            throw error;
        });
};
