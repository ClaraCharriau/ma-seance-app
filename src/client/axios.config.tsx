import axios, { AxiosResponse } from 'axios';
import { toast } from 'react-toastify';
import config from '../config/config.helper';
import { UserToken } from '../model/UserToken';

/**
 * Set axios global configuration
 */
export const axiosInstance = axios.create({
    baseURL: config.services.prefix,
    headers: {
        'Content-Type': 'application/json'
    }
});

/**
 * Set authentication header if token is found
 */
axiosInstance.interceptors.request.use(config => {
    const maSeanceIdToken = localStorage.getItem('maSeanceId');
    if (maSeanceIdToken) {
        const parsedUsertoken: UserToken = JSON.parse(maSeanceIdToken);
        config.headers.Authorization = `Bearer ${parsedUsertoken.access_token}`;
    }
    return config;
});

/**
 * Set global behavior for responses
 */
axiosInstance.interceptors.response.use(
    response => handleResponse(response),
    error => handleError(error)
);

export const handleError = (error: any) => {
    if (error.response.status === 401 || error.response.status === 403) {
        throw new Response(error.response.data, {
            status: error.response.status,
            statusText: error.response.statusText
        });
    }
    toast.error("Une erreur s'est produite.");
    if (error.response) {
        throw new Response(error.response.data, {
            status: error.response.status,
            statusText: error.response.statusText
        });
    }
};

export const handleResponse = (response: AxiosResponse) => {
    return response.data;
};
