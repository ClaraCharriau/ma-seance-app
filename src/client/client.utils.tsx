import { AxiosResponse } from 'axios';

export const HOST = 'http://localhost:7878';

export const handleError = (error: any) => {
    if (error.response) {
        throw new Response(error.response.data, {
            status: error.response.status,
            statusText: error.response.statusText
        });
    }
    console.error(error.message);
    throw error;
};

export const handleResponse = (response: AxiosResponse) => {
    return response.data;
};

export const handleDeleteResponse = (response: AxiosResponse) => {
    return response;
};
