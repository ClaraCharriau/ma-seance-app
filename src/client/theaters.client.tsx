import axios from 'axios';
import { HOST, handleError, handleResponse } from './client.utils';
import { Theater } from '../models/Theater';

const theaters_path = HOST + '/theaters';

export const getTheaterById = async (theaterId: number): Promise<Theater> => {
    return await axios
        .get(theaters_path + '/' + theaterId)
        .then(response => handleResponse(response))
        .catch(error => handleError(error));
};
