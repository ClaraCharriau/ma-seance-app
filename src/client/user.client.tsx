import { Theater } from '../models/Theater';
import axios from 'axios';
import { handleError, handleResponse } from './client.utils';

const host = 'http://localhost:7878';
const users_path = host + '/users';

export const getUserFavTheaters = async (id: number): Promise<Theater[]> => {
    return await axios
        .get(users_path + '/' + id + '/fav-theaters')
        .then(response => handleResponse(response))
        .catch(error => handleError(error, 'GET /fav-theaters'));
};
