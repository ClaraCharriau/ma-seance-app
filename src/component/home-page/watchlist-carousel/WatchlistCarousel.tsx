import { useEffect, useState } from 'react';
import { User } from '../../../model/User';
import { Movie } from '../../../model/Movie';
import { getUserFavMovies } from '../../../client/users/user.client';
import { useMediaQuery } from 'react-responsive';
import { CarouselProvider, Slide, Slider } from 'pure-react-carousel';
import style from './WatchlistCarousel.module.css';
import MovieCard from '../movie-card/MovieCard';
import SeeDetailsBtn from '../../common/see-details-btn/SeeDetailsBtn';
import EmptySection from '../../common/empty-section/EmptySection';
import Spinner from '../../common/spinner/Spinner';

interface FavMoviesCarouselProps {
    currentUser: User;
}

const WatchlistCarousel = (props: FavMoviesCarouselProps) => {
    const { currentUser } = props;
    const [favMovies, setFavMovies] = useState<Movie[]>([]);
    const [favMoviesTotal, setFavMoviesTotal] = useState<number>(0);
    const [loading, setLoading] = useState(true);
    const [error, setError] = useState('');
    const ERROR_MESSAGE = "Une erreur s'est produite lors du chargement des films.";

    useEffect(() => {
        const timeoutId = setTimeout(() => {
            setError(ERROR_MESSAGE);
            setLoading(false);
        }, 5000); // 5s

        getUserFavMovies(currentUser.id)
            .then(movies => {
                setFavMovies(movies.records);
                setFavMoviesTotal(movies._metadata.total_count);
                setLoading(false);
                clearTimeout(timeoutId);
            })
            .catch(() => {
                setError(ERROR_MESSAGE);
                setLoading(false);
            });

        return () => clearTimeout(timeoutId);
        // eslint-disable-next-line
    }, []);

    const isDesktop = useMediaQuery({
        query: '(min-device-width: 800px)'
    });
    const slideCount = isDesktop ? 2 : 4;

    return (
        <>
            <CarouselProvider
                visibleSlides={slideCount}
                totalSlides={favMovies.length}
                step={1}
                naturalSlideWidth={90}
                naturalSlideHeight={120}
                className={style.carousel}
            >
                {loading ? (
                    <Spinner />
                ) : error ? (
                    <div className={style.error}>{error}</div>
                ) : favMovies.length > 0 ? (
                    <>
                        <div className={style.sliderWrapper}>
                            <Slider className={style.slider}>
                                {favMovies.map((movie, index) => (
                                    <Slide index={index} key={movie.id} className={style.slide}>
                                        <MovieCard movie={movie} key={movie.id} />
                                    </Slide>
                                ))}
                            </Slider>
                            <div className={style.carouselShadow}></div>
                        </div>
                    </>
                ) : (
                    <EmptySection itemType="movie" />
                )}
            </CarouselProvider>
            <SeeDetailsBtn text={`Voir les ${favMoviesTotal} films >`} navigatePath="/watchlist" showIcon={false} />
        </>
    );
};

export default WatchlistCarousel;
