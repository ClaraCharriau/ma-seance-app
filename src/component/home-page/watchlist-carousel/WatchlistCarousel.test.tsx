import { act, render, waitFor } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { getUserFavMovies } from '../../../client/users/user.client';
import mockMovies from '../../../mocks/users/fav-movies.json';
import WatchlistCarousel from './WatchlistCarousel';

jest.mock('../../../client/users/user.client', () => ({
    getUserFavMovies: jest.fn()
}));
const mockGetUserFavMovies = getUserFavMovies as jest.MockedFunction<typeof getUserFavMovies>;

describe('Watchlist movie carousel component tests', () => {
    it('renders movie Carousel component', async () => {
        // Given
        const mockUser = {
            id: "1",
            pseudo: 'Jane',
            email: 'test@mail.com'
        };
        mockGetUserFavMovies.mockResolvedValue(mockMovies);

        // When
        const component = render(
            <BrowserRouter>
                <WatchlistCarousel currentUser={mockUser} />
            </BrowserRouter>
        );

        // Then
        await waitFor(() => {
            expect(component.getByAltText("Affiche du film Pauvres crÃ©atures")).toBeInTheDocument();
        });
        expect(component.baseElement).toMatchSnapshot();
    });

    it('renders error when failing to get watchlist movies', async () => {
        // Given
        const mockUser = {
            id: "1",
            pseudo: 'Jane',
            email: 'test@mail.com'
        };
        mockGetUserFavMovies.mockRejectedValue(new Error());
        let component: any;

        // When
        act(() => {
            component = render(
                <BrowserRouter>
                    <WatchlistCarousel currentUser={mockUser} />
                </BrowserRouter>
            );
        });

        // Then
        await waitFor(() => {
            expect(component.getByText("Une erreur s'est produite lors du chargement des films.")).toBeInTheDocument();
            expect(component.container).toMatchSnapshot();
        });
    });

    it('renders error when time out', async () => {
        // Given
        const mockUser = {
            id: "1",
            pseudo: 'Jane',
            email: 'test@mail.com'
        };
        let component: any;

        // When
        act(() => {
            component = render(
                <BrowserRouter>
                    <WatchlistCarousel currentUser={mockUser} />
                </BrowserRouter>
            );
        });

        // Then
        await waitFor(() => {
            setTimeout(() => {
                console.log('time out');
            }, 5001);
            expect(component.getByText("Une erreur s'est produite lors du chargement des films.")).toBeInTheDocument();
            expect(component.container).toMatchSnapshot();
        });
    });
});
