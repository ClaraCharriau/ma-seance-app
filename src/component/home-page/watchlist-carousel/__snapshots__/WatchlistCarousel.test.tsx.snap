// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Watchlist movie carousel component tests renders error when failing to get watchlist movies 1`] = `
<div>
  <div
    class="carousel carousel"
  >
    <div
      class="error"
    >
      Une erreur s'est produite lors du chargement des films.
    </div>
  </div>
</div>
`;

exports[`Watchlist movie carousel component tests renders error when time out 1`] = `
<div>
  <div
    class="carousel carousel"
  >
    <div
      class="error"
    >
      Une erreur s'est produite lors du chargement des films.
    </div>
  </div>
</div>
`;

exports[`Watchlist movie carousel component tests renders movie Carousel component 1`] = `
<body>
  <div>
    <div
      class="carousel carousel"
    >
      <div
        class="sliderWrapper"
      >
        <div
          aria-label="slider"
          aria-live="polite"
          class="horizontalSlider___281Ls carousel__slider carousel__slider--horizontal slider"
          role="listbox"
          tabindex="0"
        >
          <div
            class="carousel__slider-tray-wrapper carousel__slider-tray-wrap--horizontal"
          >
            <div
              class="sliderTray___-vHFQ sliderAnimation___300FY carousel__slider-tray carousel__slider-tray--horizontal"
              style="width: 100%; transform: translateX(0%) translateX(0px); flex-direction: row;"
            >
              <div
                aria-label="slide"
                aria-selected="true"
                class="slide___3-Nqo slideHorizontal___1NzNV carousel__slide carousel__slide--visible slide"
                role="option"
                style="width: 25%; padding-bottom: 33.333333333333336%;"
                tabindex="0"
              >
                <div
                  class="slideInner___2mfX9 carousel__inner-slide"
                >
                  <a
                    class="card"
                    href="/movies/3d8f1342-15f1-44b1-a48f-4581d654b94a/day-1"
                  >
                    <img
                      alt="Affiche du film Pauvres crÃ©atures"
                      class="movieCardImg"
                      src="https://image.tmdb.org/t/p/w1280/5fT98da9ccWN2xr8VOJrSBp3Cdw.jpg"
                    />
                  </a>
                </div>
              </div>
              <div
                aria-label="slide"
                aria-selected="true"
                class="slide___3-Nqo slideHorizontal___1NzNV carousel__slide carousel__slide--visible slide"
                role="option"
                style="width: 25%; padding-bottom: 33.333333333333336%;"
                tabindex="0"
              >
                <div
                  class="slideInner___2mfX9 carousel__inner-slide"
                >
                  <a
                    class="card"
                    href="/movies/f7e41334-179d-451a-a1d4-e1b9a60ee785/day-1"
                  >
                    <img
                      alt="Affiche du film Argylle"
                      class="movieCardImg"
                      src="https://image.tmdb.org/t/p/w1280/uVAk2YliqInQfH4B4vzZ75rwcNB.jpg"
                    />
                  </a>
                </div>
              </div>
              <div
                aria-label="slide"
                aria-selected="true"
                class="slide___3-Nqo slideHorizontal___1NzNV carousel__slide carousel__slide--visible slide"
                role="option"
                style="width: 25%; padding-bottom: 33.333333333333336%;"
                tabindex="0"
              >
                <div
                  class="slideInner___2mfX9 carousel__inner-slide"
                >
                  <a
                    class="card"
                    href="/movies/cff01b1b-a943-4bcf-a396-afd80417150a/day-1"
                  >
                    <img
                      alt="Affiche du film Dune"
                      class="movieCardImg"
                      src="https://image.tmdb.org/t/p/w1280/qpyaW4xUPeIiYA5ckg5zAZFHvsb.jpg"
                    />
                  </a>
                </div>
              </div>
              <div
                aria-label="slide"
                aria-selected="true"
                class="slide___3-Nqo slideHorizontal___1NzNV carousel__slide carousel__slide--visible slide"
                role="option"
                style="width: 25%; padding-bottom: 33.333333333333336%;"
                tabindex="0"
              >
                <div
                  class="slideInner___2mfX9 carousel__inner-slide"
                >
                  <a
                    class="card"
                    href="/movies/dd4cdbd2-44b7-4fb4-84b4-4f016d379ddb/day-1"
                  >
                    <img
                      alt="Affiche du film The Grand Budapest Hotel"
                      class="movieCardImg"
                      src="https://image.tmdb.org/t/p/w1280/atLMzzA7pOB0BdfM89V7BbdtLN6.jpg"
                    />
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="carouselShadow"
        />
      </div>
    </div>
    <a
      class="seeDetailsBtn"
      href="/watchlist"
    >
      Voir les 521 films &gt;
    </a>
  </div>
</body>
`;
